### ISSUE TITLE
[Bug] FirebaseCrashlytics | Error serializing key/value metadata.

### ISSUE THREAD

**Issue by bilck** [firebase/quickstart-unity#1090]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2020.3.13f1
  * Firebase Unity SDK version: 8.0.0
  * Source you installed the SDK: .unitypackage
  * Problematic Firebase Component: Crashlytics
  * Other Firebase Components in use: Analytics, RemoteConfig, Messaging
  * Additional SDKs you are using: Playfab, ARCore, DevToDev
  * Platform you are using the Unity editor on: Windows
  * Platform you are targeting: Android
  * Scripting Runtime: IL2CPP

### [REQUIRED] Please describe the issue here:

We are getting the following error on our Android builds:
```
Error serializing key/value metadata.
java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1441)
	at java.util.HashMap$EntryIterator.next(HashMap.java:1475)
	at java.util.HashMap$EntryIterator.next(HashMap.java:1473)
	at java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet$1.next(Collections.java:1708)
	at java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet$1.next(Collections.java:1701)
	at org.json.JSONObject.<init>(JSONObject.java:138)
	at com.google.firebase.crashlytics.internal.common.p.a(Unknown Source:2)
	at com.google.firebase.crashlytics.internal.common.p.a(Unknown Source:7)
	at com.google.firebase.crashlytics.internal.common.CrashlyticsController$9.a(Unknown Source:19)
	at com.google.firebase.crashlytics.internal.common.CrashlyticsController$9.call(Unknown Source:0)
	at com.google.firebase.crashlytics.internal.common.CrashlyticsBackgroundWorker$3.then(Unknown Source:2)
	at com.google.android.gms.tasks.h.run(Unknown Source:26)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at com.google.firebase.crashlytics.internal.common.l$1$1.a(Unknown Source:2)
	at com.google.firebase.crashlytics.internal.common.b.run(Unknown Source:5)
	at java.lang.Thread.run(Thread.java:923)
```

#### Steps to reproduce:
Just open the app and this exception will be logged.

We don't know exactly which CustomKey is triggering this issue. It would be very helpful to log which Key had issues when being serialized.

We double-checked and no Keys or values exceed 1024 characters.

**Comment by paulinon** [comment]
Hi @bilck,

In order to identify what's causing this behavior, could you provide a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of your project along with the steps taken before facing this issue?

**Comment by bilck** [comment]
HI @paulinon,

Unfortunately right now I can't work on reproducing this example, **but** I can provide you with a development APK or access to our repository. Would that work for you?

**Comment by paulinon** [comment]
Hi @bilck,

That would be great. Make sure you also include the steps taken before facing this issue.

**Comment by bilck** [comment]
Is an development APK enough to investigate the issue?

This exception seems that it is caused by multiple metadata being set at the same time.
Would setting 10+ metadata at the same time be an issue? The Unity Firebase SDK only supports setting one custom data at a time (`Crashlytics.SetCustomKey`).

**Comment by paulinon** [comment]
Hi @blick,

Between the two, I prefer having access to your repository so that I can try to replicate your issue and hopefully narrow down what's causing this behavior.

In order to verify if the issue persists elsewhere, could you try using the quickstart and see if you get the same results?

**Comment by bilck** [comment]
@paulinon I guess that the issue happens when trying to send multiple Custom Data at the same time.

To add you to the repository I need your BitBucket user.

**Comment by paulinon** [comment]
Hi @bilck,

It would be great if you could upload your repository here on GitHub and then add me as a collaborator so that I can inspect your project.

Also, have you gotten the same results when using the quickstart?

**Comment by bilck** [comment]
I can't push the repository to Github due to some large files (over 100MB) in the repository history.
Unfortunately that's not something I'm willing to fix at the moment.
I can add an user to the Bitbucket repository though.

**Comment by paulinon** [comment]
Hi @bilck,

In the meantime, could you verify if this behavior persists when setting a lot of custom metadata concurrently in the quickstart?

**Comment by vimanyu** [comment]
Thanks for all the info on this.  We have reached out to Crashlytics android team for suggestions and will have an update soon.

**Comment by samedson** [comment]
Hey folks, sorry for the Exception here. We've fixed this in the underlying Android SDK: https://github.com/firebase/firebase-android-sdk/pull/2851

This will be fixed in the next Unity SDK release.

If you need to pull in the new Android SDK ASAP, you can modify your gradle config to point to a local snapshot and build the Crashlytics Android SDK directly: https://github.com/firebase/firebase-android-sdk#commands

**Comment by bilck** [comment]
> Hi @bilck,
> 
> In the meantime, could you verify if this behavior persists when setting a lot of custom metadata concurrently in the quickstart?

Yeah, that's when this issue happens.
I'm glad this issue was found and fixed, sorry for not being able to give you access to the repository for a quicker investigation.
:)

**Comment by DellaBitta** [comment]
This has been fixed in the Crashlytics SDK which will be pulled in as a dependency in the next Unity release `8.2.0`.  I'm going to close this issue for now but if the issue is not resolved by the upcoming 8.2.0 release then please reopen the issue here.  Thanks!
